### PR TITLE
Dont send content-length on 1XX or 204 responses

### PR DESCRIPTION
**Pull request by jakul**
Fixes https://github.com/pallets/werkzeug/issues/1025

### CODE REVIEW

**Comment by jakul** [comment]
The failed test doesn't seem to have anything to do with me:

```
___________________ TestMemcachedCache.test_generic_get_dict ___________________
self = <tests.contrib.test_cache.TestMemcachedCache object at 0x1076dd550>
c = <werkzeug.contrib.cache.MemcachedCache object at 0x1076da630>
    def test_generic_get_dict(self, c):
>       assert c.set('a', 'a')
E       assert 0
E        +  where 0 = <bound method MemcachedCache.set of <werkzeug.contrib.cache.MemcachedCache object at 0x1076da630>>('a', 'a')
E        +    where <bound method MemcachedCache.set of <werkzeug.contrib.cache.MemcachedCache object at 0x1076da630>> = <werkzeug.contrib.cache.MemcachedCache object at 0x1076da630>.set
tests/contrib/test_cache.py:60: AssertionError
---------------------------- Captured stdout setup -----------------------------
/Users/travis/build/pallets/werkzeug/.xprocess/memcached$ memcached
process 'memcached' started pid=4020
memcached process startup detected
----------------------------- Captured stderr call -----------------------------
/Users/travis/build/pallets/werkzeug/.tox/py-normal/lib/python3.5/site-packages/memcache.py:1391: ResourceWarning: unclosed <socket.socket fd=19, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 50579)>
  if self._get_socket():
============== 1 failed, 528 passed, 17 skipped in 125.61 seconds ==============
```

**Comment by mitsuhiko** [comment]
That seems to make sense. Did this cause any issues in practice?

File: werkzeug/wrappers.py [hunk 1]
```diff
@@ -1181,7 +1176,8 @@ def get_wsgi_headers(self, environ):
         # the response.  We however should not do that if we have a 304
         # response.
         if self.automatically_set_content_length and \
-           self.is_sequence and content_length is None and status != 304:
+           self.is_sequence and content_length is None and \
+           status not in (204, 304) and 200 >= status > 100:
```
Comment (untitaker) [inline werkzeug/wrappers.py]: Don't we want content-length determination for error pages (404 ,500) as well?

**Comment by jakul** [comment]
@mitsuhiko I haven't seen any actual problems caused by this. I have been investigating a few [other bugs](https://github.com/apiaryio/dredd/issues/196) to do with HTTP 204 statuses, which is when I noticed the problem.

@untitaker Good spot! I'll fix it now

File: werkzeug/wrappers.py (hunk 1 above)
Comment (jakul) [inline werkzeug/wrappers.py]: Good point. I'm pushing a fix now.

**Comment by lepture** [comment]
@jakul could you rebase your pr?

@untitaker Is is ready to merge this pr?

**Comment by jakul** [comment]
@lepture rebased

**Comment by untitaker** [comment]
After the last comment from @jakul no fix has been pushed, which is why this was on ice.

Tests seem to pass, there's one ever-flaky memcached test, but that's irrelevant.

**Comment by mitsuhiko** [comment]
Applied this now in 1c70c3e1

**Pull request closed without merging**